src/NuGet.Jobs.Catalog2Registration/Schema/RegistrationUrlBuilder.cs

```diff
@@ -0,0 +1,110 @@
+            _gzippedBaseUrl = EnsureTrailingSlash(_options.Value.GzippedBaseUrl);
+            _semver2BaseUrl = EnsureTrailingSlash(_options.Value.SemVer2BaseUrl);
+        }
+
+        public string GetIndexPath(string id)
+        {
+            return $"{Uri.EscapeUriString(id.ToLowerInvariant())}/index.json";
+        }
+
+        public string GetIndexUrl(HiveType hive, string id)
+        {
+            return GetBaseUrl(hive) + GetIndexPath(id);
+        }
+
+        public string GetInlinedPageUrl(HiveType hive, string id, NuGetVersion lower, NuGetVersion upper)
+        {
+            return GetIndexUrl(hive, id) + "#" + GetPageFragment(lower, upper);
+        }
+
+        public string ConvertHive(HiveType fromHive, HiveType toHive, string url)
+        {
+            if (url == null)
+            {
+                return null;
+            }
+
+            var path = ConvertToPath(fromHive, url);
+            return GetBaseUrl(toHive) + path;
+        }
+
+        public string ConvertToPath(HiveType hive, string url)
+        {
+            var baseUrl = GetBaseUrl(hive);
+            Guard.Assert(url.StartsWith(baseUrl), $"URL '{url}' does not start with expected base URL '{baseUrl}'.");
+            return url.Substring(baseUrl.Length);
+        }
+
+        public string GetPageUrl(HiveType hive, string id, NuGetVersion lower, NuGetVersion upper)
+        {
+            return GetBaseUrl(hive) + GetPagePath(id, lower, upper);
+        }
+
+        public string GetPagePath(string id, NuGetVersion lower, NuGetVersion upper)
+        {
+            return $"{Uri.EscapeUriString(id.ToLowerInvariant())}/{GetPageFragment(lower, upper)}.json";
+        }
+
+        private static string GetPageFragment(NuGetVersion lower, NuGetVersion upper)
+        {
+            return
+                $"page/" +
+                $"{Uri.EscapeUriString(lower.ToNormalizedString().ToLowerInvariant())}/" +
+                $"{Uri.EscapeUriString(upper.ToNormalizedString().ToLowerInvariant())}";
+        }
+
+        public string GetLeafPath(string id, NuGetVersion version)
+        {
+            return
+                $"{Uri.EscapeUriString(id.ToLowerInvariant())}/" +
+                $"{Uri.EscapeUriString(version.ToNormalizedString().ToLowerInvariant())}.json";
+        }
+
+        public string GetLeafUrl(HiveType hive, string id, NuGetVersion version)
+        {
+            return GetBaseUrl(hive) + GetLeafPath(id, version);
+        }
+
+        private string GetBaseUrl(HiveType hive)
+        {
+            switch (hive)
+            {
+                case HiveType.Legacy:
+                    return _legacyBaseUrl;
+                case HiveType.Gzipped:
+                    return _gzippedBaseUrl;
+                case HiveType.SemVer2:
+                    return _semver2BaseUrl;
+                default:
+                    throw new NotImplementedException($"The hive type '{hive}' does not have configured base URL.");
```

Comment:
nit: `The hive type '{hive}' does not have *a* configured base URL.`

Reply:
Oh nice catch.